marshal/admin.go

```diff
@@ -0,0 +1,322 @@
+type claimAttempt struct {
+	topic  string
+	partID int
+	// What we'd like to set the offset of a particular partition to be.
+	newOffset int64
+	// What the current offset of a particular partition is.
+	currentOffset int64
+}
+
+// addClaimAttempt adds a successfully-claimed partition to our Admin.
+func (a *consumerGroupAdmin) addClaimAttempt(topic string, partID int, currentOffset, newOffset int64) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	c := claimAttempt{
+		topic:         topic,
+		partID:        partID,
+		newOffset:     newOffset,
+		currentOffset: currentOffset}
+	a.claims = append(a.claims, c)
+}
+
+// claimHealth returns whether or not any of the admin's claims have failed to heartbeat.
+func (a *consumerGroupAdmin) claimsHealthy() bool {
+	return atomic.LoadInt32(a.claimHealth) == 0
+}
+
+func (m *Marshaler) NewAdmin(groupID string, pauseTimeout time.Duration) (Admin, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	if len(m.consumers) != 0 {
+		return nil, fmt.Errorf(
+			"The Marshaler instance bound to an Admin should not have any consumers.")
+	}
+	return &consumerGroupAdmin{
+		clientID:     m.clientID,
+		groupID:      groupID,
+		marshaler:    m,
+		pauseTimeout: pauseTimeout,
+
+		claimHealth: new(int32),
+		lock:        &sync.RWMutex{},
+		// Used for testing.
+		steps: new(int32),
+	}, nil
+}
+
+// release releases claim on topic, partition. Optionally resets the offset on the partition.
+func (a consumerGroupAdmin) release(c claimAttempt, withNewOffset bool) bool {
+	releaseOffset := c.currentOffset
+	if withNewOffset {
+		if err := a.marshaler.Heartbeat(c.topic, c.partID, c.newOffset); err != nil {
+			log.Errorf("[%s:%d] Admin failed to heartbeat", c.topic, c.partID)
+			return false
+		}
+		releaseOffset = c.newOffset
+	}
+
+	if err := a.marshaler.ReleasePartition(c.topic, c.partID, releaseOffset); err != nil {
+		log.Errorf("[%s:%d] Admin failed to release partition with offset %d: %s",
+			c.topic, c.partID, releaseOffset, err)
+		return false
+	}
+	return true
+}
+
+// releaseClaims releases all claims the Admin has, optionally resetting their offsets.
+func (a *consumerGroupAdmin) releaseClaims(newOffset bool) error {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+
+	if !newOffset {
+		log.Infof("Admin releasing claims without resetting offsets.")
+	}
+
+	results := make(chan bool)
+	defer close(results)
+	for _, claimAttempt := range a.claims {
+
+		// Update our step counter. Releasing with a new offset will Heartbeat then ReleasePartition,
+		// releasing without resetting the offset will just ReleasePartition
+		atomic.AddInt32(a.steps, 1)
+		if newOffset {
+			atomic.AddInt32(a.steps, 1)
+		}
+
+		go func() {
+			results <- a.release(claimAttempt, newOffset)
+		}()
+	}
+
+	numResults := len(a.claims)
+	for numResults > 0 {
+		select {
+		case result := <-results:
+			if !result {
+				return fmt.Errorf("Admin failed to set a Kafka offset. Aborting!")
+			}
+			numResults--
+		default:
```

Comment:
This is a busy loop. If the channel is empty the select will exit immediately and the for will just loop causing it to basically `while true {}` until it's done. I think I agree that firing up a WaitGroup would be better here since you know how many claims you have a priori.

You can also put a timer as the second part of the select. That way it's fast if tehre's anything in channel but else it will block for 100ms or something.